Lib9c/Action/Buy.cs

```diff
@@ -148,159 +269,174 @@ public override IAccountStateDelta Execute(IActionContext context)
                 throw new NotEnoughClearedStageLevelException(addressesHex, GameConfig.RequireClearedStageLevel.ActionsInShop, current);
             }
 
-            if (!states.TryGetState(ShopState.Address, out Bencodex.Types.Dictionary shopStateDict))
+            List<PurchaseResult> purchaseResults = new List<PurchaseResult>();
+            List<SellerResult> sellerResults = new List<SellerResult>();
+            MaterialItemSheet materialSheet = states.GetSheet<MaterialItemSheet>();
+            buyerMultipleResult = new BuyerMultipleResult();
+            sellerMultipleResult = new SellerMultipleResult();
+
+            foreach (var purchaseInfo in purchaseInfos)
             {
-                throw new FailedLoadStateException($"{addressesHex}Aborted as the shop state was failed to load.");
-            }
+                PurchaseResult purchaseResult = new PurchaseResult(purchaseInfo.productId);
+                Address shardedShopAddress =
+                    ShardedShopState.DeriveAddress(purchaseInfo.itemSubType, purchaseInfo.productId);
+                Address sellerAgentAddress = purchaseInfo.sellerAgentAddress;
+                Address sellerAvatarAddress = purchaseInfo.sellerAvatarAddress;
+
+                purchaseResults.Add(purchaseResult);
+                if (!states.TryGetState(shardedShopAddress, out Bencodex.Types.Dictionary shopStateDict))
+                {
+                    purchaseResult.errorCode = ERROR_CODE_FAILED_LOADING_STATE;
+                    continue;
+                }
+                sw.Stop();
+                Log.Verbose("{AddressesHex}Buy Get ShopState: {Elapsed}", addressesHex, sw.Elapsed);
```

Comment:
This seems to be unmeasured on looping.

Reply:
I think it is correct to loop because the address varies depending on the item type and itemId.